src/features/defi/providers/thorchain-savers/components/ThorchainSaversManager/Deposit/components/Approve.tsx

```diff
@@ -0,0 +1,355 @@
+            spender: saversRouterContractAddress,
+            from: fromAccountId(accountId).account,
+            chainId: asset.chainId,
+          }),
+        validate: (allowanceCryptoBaseUnit: string) => {
+          return bnOrZero(allowanceCryptoBaseUnit).gte(amountCryptoBaseUnit)
+        },
+        interval: 15000,
+        maxAttempts: 60,
+      })
+
+      const estimatedDepositGasCryptoPrecision = await (async () => {
+        const maybeQuote = await getMaybeThorchainSaversDepositQuote({
+          asset,
+          amountCryptoBaseUnit,
+        })
+        if (maybeQuote.isErr()) throw new Error(maybeQuote.unwrapErr())
+        const quote = maybeQuote.unwrap()
+        const thorContract = getOrCreateContractByType({
+          address: saversRouterContractAddress!,
+          type: ContractType.ThorRouter,
+          chainId: asset.chainId,
+        })
+
+        const data = thorContract.interface.encodeFunctionData('depositWithExpiry', [
+          quote.inbound_address,
+          fromAssetId(assetId).assetReference,
+          amountCryptoBaseUnit.toString(),
+          quote.memo,
+          quote.expiry,
+        ])
+
+        const adapter = chainAdapterManager.get(asset.chainId) as unknown as EvmChainAdapter
+
+        const customTxInput = await createBuildCustomTxInput({
+          accountNumber,
+          adapter,
+          data,
+          value: '0', // this is not a token send, but a smart contract call so we don't send anything here, THOR router does
+          to: saversRouterContractAddress!,
+          wallet,
+        })
+
+        const fees = await adapter.getFeeData({
+          to: customTxInput.to,
+          value: customTxInput.value,
+          chainSpecific: {
+            from: fromAccountId(accountId).account,
+            data: customTxInput.data,
+          },
+        })
+
+        const fastFeeCryptoBaseUnit = fees.fast.txFee
+        const fastFeeCryptoPrecision = bnOrZero(
+          bn(fastFeeCryptoBaseUnit).div(bn(10).pow(feeAsset.precision)),
+        )
+
+        return fastFeeCryptoPrecision.toString()
+      })()
+
+      dispatch({
+        type: ThorchainSaversDepositActionType.SET_DEPOSIT,
+        payload: { estimatedGasCryptoPrecision: estimatedDepositGasCryptoPrecision },
+      })
+
+      trackOpportunityEvent(
+        MixPanelEvents.DepositApprove,
+        {
+          opportunity: opportunityData,
+          fiatAmounts: [state.deposit.fiatAmount],
+          cryptoAmounts: [{ assetId, amountCryptoHuman: state.deposit.cryptoAmount }],
+        },
+        assets,
+      )
+
+      onNext(DefiStep.Confirm)
+      dispatch({ type: ThorchainSaversDepositActionType.SET_LOADING, payload: false })
+    } catch (error) {
+      console.error(error)
+      toast({
+        position: 'top-right',
+        description: translate('common.somethingWentWrongBody'),
+        title: translate('common.somethingWentWrong'),
+        status: 'error',
+      })
```

Comment:
there's also this `const { showErrorToast } = useErrorHandler()` so you can do `showErrorToast(error)` with error logging

Reply:
heck yeah 547805a86e